KEP-4819: Container log rotation on Disk perssure (#5022)
kubernetes/enhancements: keps/sig-node/5032-log-eviction-on-disk-pressure/README.md

### Risks and Mitigations

Do not see any risk as of now.

## Design Details

1. Implement a new function (splitAndRotateLatestLog) to be called by rotateLog function (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/logs/container_log_manager.go#L235-L257)
2. The rotateLog is being called by each worker for the container assigned to it.
3. It does cleanup to delete all non-rotated (rotated logs have time suffix in name) and .tmp files generated (and if not deleted) in last log rotation.
4. Then it rotates un rotated files (it does not rotate the active file) and deletes oldest rotated files till containerLogMaxFiles-2 files are left. This is because the non rotated active file be rotated and new active file will be created. Which will add upto containerLogMaxFiles.
5. Before doing this exact above step, in the new design, it will split the large active log file in size of containerLogMaxSize and name them <active-log-file-name>.part<i>.

> 6. Let's say it created n parts, after this, it can do rotate of n-1 parts, keeping last nth part active and do delete. (Basically step 4)

Reviewer: > Adding @leonzz to help review the potential impact on the log collection pipeline.

> keeping last nth part active

If you truncate a 11M `container.log` in place to 1M and copy out the first 10M to `container.part1`,  it could potentially mess up the read position with some log collectors. For example fluentbit would re-read the 1M content after the truncation, resulting in duplicated logs, because it would always read from beginning when truncation is detected (see the discussion in https://github.com/fluent/fluent-bit/issues/375#issuecomment-363040116)

For log collectors to behave correctly, it's better to just create a new empty active file but this means the nth part of the last active file could be much smaller than containerLogMaxSize

Author: Could we think about keeping the tailing logs to the last part itself till the size is about containerLogMaxSize/2.
Create a new part for tailing logs only if it more than containerLogMaxSize/2.

And create entirely new active file always.

Author: @leonzz 
Can we think about keeping the tailing logs to the last part itself till the size is about containerLogMaxSize/2.
Create a new part for tailing logs only if it is more than containerLogMaxSize/2.

And create entirely new active file always.

Reviewer: > Create a new part for tailing logs only if it is more than containerLogMaxSize/2.

To me this sounds like a reasonable tradeoff to `always create new active files`.

So we could have files sizes anywhere in between 0.5 to 1.5 times containerLogMaxSize, which I assume can still resolve the problems you encounter with super large files?

    - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
    - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

Split, Clean and Rotate container logs to avoid disk pressure on kubelet host.

## Motivation

> - We manage kubernetes ecosystem at our organization. A lot of our kubelet hosts experienced Disk pressure as a certain set of pods was generating very high logs. The rate was around 3-4Gib logs generated in 15 minutes. We had kubelet configs containerLogMaxSize set to 200Mib and containerLogMaxFiles set to 6. But the .gz files (tar log files of pods) were of size around 500-600Gib. We observed that container log rotation was slow for us.

Reviewer 2: This is very specific and doesn't represent Kubernetes as an organization. @Zeel-Patel you should reword this motivation to cover the _Kubernetes project_ motivation to make this change.

### What would you like to be added?

> We expect that the log file size is always under set kubelet config limit (i.e., containerLogMaxSize) which can help such disk pressure issues in the future.

Reviewer 2: In this paragraph, who does “we“ refer to?

### Why is this needed?

It often happens that the containers generating heavy log data have compressed log file with size exceeding the containerLogMaxSize limit set in kubelet config.

For example, kubelet has
```
containerLogMaxSize = 200M
containerLogMaxFiles = 6
```

> ### Spec 1
> 
> Continuously generating 10Mib with 0.1 sec sleep in between

Reviewer 2: ```suggestion
### Example 1

This manifest defines a Job that continuously generates 10MiB of logs with 0.1 sec sleep in between (100MiB/s):
```

-rw-r--r-- 1 root root 183336376 Jan  1 11:29 0.log.20250101-112616.gz
-rw-r----- 1 root root 205360966 Jan  1 11:29 0.log.20250101-112911
```

If the pod had been generating logs in Gigabytes with minimal delay, it can cause disk pressure on kubelet host and that can affect other pods running in the same kubelet.

### Goals

- There is a ContainerLogManager in every kubelet. It runs an infinite go routine and checks active log file(on which all container log read write happens) size. If that exceeds the above mentioned limit (containerLogMaxSize), it starts parallel workers. Each worker creates a tar of the active file, deletes tars till there are containerLogMaxFiles files in total and creates a new active file for container.
- Goal is to Split large active log file in size containerLogMaxSize, and then do the rest of the operations done by containerLogManager.

## Proposal
- The container log rotation working now shpuld work as is, but it will ensure that before rotating file, it is under the size limit set. This way, every tar present for a container under host will surely be under containerLogMaxSize. This can avoid disk pressure on the host.

### Risks and Mitigations

> Do not see any risk as of now.

Reviewer 2: Are you sure?

## Design Details

> 1. Implement a new function (splitAndRotateLatestLog) to be called by rotateLog function (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/logs/container_log_manager.go#L313-L346)

Reviewer 2: ```suggestion
1. Implement a new function (`splitAndRotateLatestLog()`) to be called by the `rotateLog` function (https://github.com/kubernetes/kubernetes/blob/9e555875e79f60a238952e91e91f79f32c053f9c/pkg/kubelet/logs/container_log_manager.go#L313-L346)
2. ```

- `<package>`: `<date>` - `<test coverage>`

##### Integration tests
- Scenarios will be covered in e2e tests. 

##### e2e tests
- Add test under `kubernetes/test/e2e_node`.
- Set low value for `containerLogMaxSize` and `containerLogMaxFiles`.
- Create a pod with generating heavy logs and expect the container's combined log size to be within `containerLogMaxSize`*`containerLogMaxFiles`.

### Graduation Criteria

**Note:** *Not required until targeted at a release.*

###### How can this feature be enabled / disabled in a live cluster?

- [X] Other
    - Describe the mechanism:
    - Will enabling / disabling the feature require downtime of the control
      plane? Yes (kubelet restart)
    - Will enabling / disabling the feature require downtime or reprovisioning
      of a node? No, restart of kubelet with updated configurations and version should work.

###### Does enabling the feature change any default behavior?
No

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
Yes

###### What happens if we reenable the feature if it was previously rolled back?

###### Are there any tests for feature enablement/disablement?

> Add UTs.

Reviewer 2: ```suggestion

- Unit tests will be added for alpha

```

?
